### ISSUE TITLE
Leaving a robot on next to a powered charger drains way too much power

### ISSUE THREAD

**Issue by black3agl3** [MightyPirates/OpenComputers#348]
I set up an ic2 mfsu fully charged, feeding power into a charger. The charger is powered with a redstone torch right next to it.

Leaving a robot turned on next to the charger drains, from what I can tell, 50 000 EU per second. If the robot is off, no power is used from the MFSU.

I am using OpenComputers 1.3.0.197 beta 1 and ic2 experimental 2.1.478.
## 

I also tried two test case programs (the robot will move forward and will be right next to the charger)

Program1:
robot.forward()
robot.back()

Program2:
robot.forward()
os.sleep(1)
robot.back()

In both situations, it seems that the same amount of power is drained from the mfsu but the power charged into the robot is less in Program 1.

**Comment by Kilobyte22** [comment]
How much energy does a computer drain? Does it have the same issue?

EDIT: Checked code for the charger, i don't see any obvious errors

**Comment by black3agl3** [comment]
I didn't test for computers. I can try that if you want.

OP = open computer power
EU = IC2 EU
Oh and ignore the number in the above post (50 000) that was just from looking at the numbers flying by. This time I actually wrote programs that would help measure the exact power usage.

Experiment 1:
I programmed a robot to turn on a charger for 200 seconds. The charger is connected to an initially fully charged mfsu. At the end of the 200 seconds, I found that the EU consumed was 1236992.

I then programmed the same robot to measure its energy use for 200 seconds. This time it is not connected to a charger. Result:160.5716 OP

EU:OP ratio = 7703

Experiment 2
The next experiment is using the same robot, almost completely discharged. I wrote a program which would charge the robot to around 90% turn off the charger, then calculate how much OP it recharged itself. Since the charger is connected to an initially full mfsu I knew how much EU was consumed.
EU consumed = 413696
OP recharged = 40386.08
EU:OP ratio = 10.2

Power usage is around 752 times more when the robot (which is on) is connected to a powered charger.

I'm not sure how you power conversion works but it seems that your charger takes the some amount of EU/t, converts the whole lot to your type of power, tries to charge the robot and then destroys any excess power.

**Comment by Kilobyte22** [comment]
In fact, the charger has an internal buffer. It takes the energy from there, puts it into the robot and puts any excess back. I will do some debugging and report back my results. Actually, don't have IC2 in my dev env. The question now is: Does that only apply to IC2 energy or also other kinds of energy

**Comment by fnuecke** [comment]
So I did some debugging, and it turns out there was an error in the IC2 implementation (it's the _only_ power system that expects the amount of energy not consumed to be returned, as opposed to the amount of energy consumed), **but** this doesn't fix anything, because apparently IC2 doesn't do anything with the returned value (otherwise this would have been obvious in the tests I did up to this point). Even though the API a) indicates it b) states in the Javadoc.

So basically what's happening is this: per tick IC2 gives me 2048 EU, I can only consume a small part of that (e.g. when charging a 10k storage battery, which was my test case, 500), return the rest (1548 EU in my test). And those go poof. Ergo way higher consumption than when IC2 gives OC less or equal of what can actually go into the internal buffers.

There was some workaround implementation for other reasons, initially (the doc recommends not returning anything for performance reasons), but it seemed unnecessarily complex so I stripped it out. Might have to look into bringing that back... though that has some other issues. Oh well.
